python/ray/autoscaler/_private/resource_demand_scheduler.py

```diff
@@ -58,27 +58,27 @@ def get_nodes_to_launch(self, nodes: List[NodeID],
             resource_demands: Vector of resource demands from the scheduler.
             usage_by_ip: Mapping from ip to available resources.
         """
-
-        if resource_demands is None:
-            logger.info("No resource demands")
-            return []
-
         node_resources, node_type_counts = self.calculate_node_resources(
             nodes, pending_nodes, usage_by_ip)
-
-        logger.info("Cluster resources: {}".format(node_resources))
-        logger.info("Node counts: {}".format(node_type_counts))
-
-        unfulfilled = get_bin_pack_residual(node_resources, resource_demands)
-        logger.info("Resource demands: {}".format(resource_demands))
-        logger.info("Unfulfilled demands: {}".format(unfulfilled))
-
-        nodes = get_nodes_for(
-            self.node_types, node_type_counts,
-            self.max_workers - len(nodes) - sum(pending_nodes.values()),
-            unfulfilled)
-        logger.info("Node requests: {}".format(nodes))
-        return nodes
+        nodes_to_add = []
+        if resource_demands is not None:
+            logger.info("Cluster resources: {}".format(node_resources))
+            logger.info("Node counts: {}".format(node_type_counts))
+            unfulfilled = get_bin_pack_residual(node_resources,
+                                                resource_demands)
+            logger.info("Resource demands: {}".format(resource_demands))
+            logger.info("Unfulfilled demands: {}".format(unfulfilled))
+
+            nodes_to_add = get_nodes_for(
+                self.node_types, node_type_counts,
+                self.max_workers - len(nodes) - sum(pending_nodes.values()),
+                unfulfilled)
+
+        nodes_to_add = _add_min_workers_nodes(self.node_types,
```

Comment:
Shouldn't we do this before calling `get_bin_pack_residual`?

Reply:
Hmm, why?

Reply:
Consider a 1 GPU request (on an idle cluster) in which `p2.8xlarge` has `min_workers: 1`. Right now, I think this will spin up a `p2.xlarge` in addition to that empty `2.8xlarge` (I think there are less contrived examples of this too). 

In general, I think this version is more likely to lead to unnecessary idle instances.

Reply:
Why do you think it will do it? _add_min_workers_nodes only updates the `nodes_to_add` to `min_workers`, basically takes the `get_bin_pack_residual` output and compares it to min_workers, if it is higher it leaves it as is, if it is lower it returns min_workers.

Reply:
Perhaps the function name is confusing, it should maybe be `_update_min_workers_constraint`

Reply:
because `get_node_for` doesn't know about this min workers requirement, so it will pick a smaller instance (the `p2.xlarge` even though the bigger instance (the `p2.8xlarge`) has to be allocated anyways.

Reply:
I see, good point.

Reply:
Good catch @wuisawesome !